### ISSUE TITLE
Adaptive zoom

### ISSUE THREAD

**Issue by net4u** [jmelosegui/GooglemapMvc#28]
Hi,
Is possible to adjust the zoom in respect to dataset available (e.g. if the coordinates in the data set are around the city centre to have e.g. zoom 15, if are around the city zoom 12, if are country wide zoom 7 etc.) 
I have a stored procedure that filter the records in a table based on a radius around a given coordinate (e.g. current position) but I don't know how to use it in MVC. (I am new to MVC and I have a lot of things that are not quite clear for me now, either I can't figure out how to bind this to a given model for plotting markers.
Thank you

**Comment by jmelosegui** [comment]
@net4u this is something you can accomplish with the current version.

in your controller before passing the marker collection to the view you can calculate the zoom value  based on whatever logic you want and then set that value to the map using the Html Helper.

```
    public partial class MarkerController
    {
       public ActionResult DataBindingToModel()
        {
           //Go to the database for the marker collection
           List<MyMarkerType> markerCollection = DataContext.GetMarkers();
           //Pass the collection to a method to calculate the best zoom
           int myZoom = CalculateZoomBasedOnMyCustomLogic(markerCollection);

           var model = new MyModel
           {
                 zoom = myZoom,
                 markers = markerCollection
           }
            
            return View(model);
        }
    }
```

then in the view just set the zoom and the marker collection. 

```
@( Html.Googlemap()
        .Name("map")
        .Zoom(Model.zoom)
        .BindTo<MyMarkerType, Marker>
        (Model.markers, mappings => 
        {
            mappings.For<MyMarkerType>(binding => binding
                    .ItemDataBound(marker, obj) =>
                    {
                        marker.Latitude = obj.Latitude;
                        marker.Longitude = obj.Longitude;
                        //more properties to map here
                       ....
                    })
            );
        })
)
```
For more info about the databinding  look [this](http://www.jmelosegui.com/map/marker/DataBindingToModel) example.

**Comment by net4u** [comment]
Hi
Thank you for reply. I think that my question, have two aspects: one is objective and refers to the code itself, and another is subjective and refers to zoom approximation for a given viewport (eg our coordinates can and scattered within a radius of 10 km and it involves a zoom level 15 or can be scattered within a radius of 200 km and it involves a zoom level 7 or 5 for example, but the assessment is subjective, I do not know a correlation between a given viewport, say 1000 pixels and a radius on the map 100 km).
In terms of binding to an existing model (with filtering ability controller) where to find some more information? As I said I have a database stored procedure to achieve filtering records in a table, with the parameters current position and distance (in km) but do not know how to use it in MVC.
My models can be simplified to the maximum so: model Jobs include IdJob, Title, Description, Country, Region, City, Address, ZIP, Latitude, Longitude and model Candidates include IdCandidate, Title, Description, Country, Region, City, Address, ZIP , Latitude, Longitude
Thank you again.

**Comment by net4u** [comment]
I try a MVC app from scratch, and I have the following:
Model for Jobs in Jobs.cs:
namespace GMap.Models
{
    using System;
    using System.Collections.Generic;
    
    public partial class Jobs
    {
        public int IdJob { get; set; }
        public Nullable<int> IdWho { get; set; }
        public string User { get; set; }
        public Nullable<System.DateTime> DateAdd { get; set; }
        public string ImageSRC { get; set; }
        public string Titlu { get; set; }
        public string WhereWhat { get; set; }
        public string Description { get; set; }
        public string Content { get; set; }
        public string Solicitare { get; set; }
        public Nullable<System.DateTime> DateStart { get; set; }
        public Nullable<System.DateTime> DateEnd { get; set; }
        public string Orar { get; set; }
        public string Localitate { get; set; }
        public string Address { get; set; }
        public Nullable<double> Latitudine { get; set; }
        public Nullable<double> Longitudine { get; set; }
        public string ZIP { get; set; }
        public string Region { get; set; }
    
        public virtual AspNetUsers AspNetUsers { get; set; }
    }
}

Controller for Jobs in JobsController.cs:

using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Entity;
using System.Linq;
using System.Net;
using System.Web;
using System.Web.Mvc;
using GMap.Models;

namespace GMap.Controllers
{
    public class JobsController : Controller
    {
        private babysitterEntities db = new babysitterEntities();

        // GET: Jobs
        public ActionResult Index()
        {
            var jobs = db.Jobs.Include(j => j.AspNetUsers);
            return View(jobs.ToList());
        }
       @*other actions*@
        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                db.Dispose();
            }
            base.Dispose(disposing);
        }
    }
}

View for Jobs in index.html:

@using System.Drawing
@using Jmelosegui.Mvc.Googlemap.Overlays

@model IEnumerable<GMap.Models.Jobs>

@{
    ViewBag.Title = "Index";
}

<h2>Index</h2>

@{ Html.GoogleMap()
                .Name("map")
                .Height(600)
                .Center(c => c.Latitude(40)
                              .Longitude(-3))
                .BindTo<Jobs, Marker>
                (Model, mappings => mappings.For<Jobs>
                        (
                            binding => binding.ItemDataBound
                            (
                                (marker, regionInfo) =>
                                {
                                    marker.Latitude = regionInfo.Latitude;
                                    marker.Longitude = regionInfo.Longitude;
                                    marker.Title = regionInfo.Title;
                                    marker.zIndex = regionInfo.zIndex;
                                    marker.Icon = new MarkerImage(String.Format("~/Images/flags/{0}", regionInfo.ImagePath)
                                                                    , new Size(18, 12)
                                                                    , new Point(0, 0)
                                                                    , new Point(0, 12));

                                    marker.Window = new InfoWindow(marker)
                                                    {
                                                        Template =
                                                        {
                                                            Content = () => ViewContext.Writer.Write(regionInfo.InfoWindowContent)
                                                        }
                                                    };
                                }
                            )
                        )
                ).Render();
}

In project I referenced Jmelosegui.Mvc.Googlemap
As expected I have at least this error when I run application:
 'System.Web.Mvc.HtmlHelper<System.Collections.Generic.IEnumerable<GMap.Models.Jobs>>' does not contain a definition for 'GoogleMap' and no extension method 'GoogleMap' accepting a first argument of type 'System.Web.Mvc.HtmlHelper<System.Collections.Generic.IEnumerable<GMap.Models.Jobs>>' could be found (are you missing a using directive or an assembly reference?)


I tried to avoid as much of your code because it is an example, records are simulated and I am interested to understand how to use the object with records in my table.

**Comment by net4u** [comment]
I assume that my return View(jobs.ToList()); is similar with your return View(DataContext.GetRegions());

**Comment by jmelosegui** [comment]
To fix the error you are facing, just add the following using to your page.

```
@using Jmelosegui.Mvc.Googlemap
```

I regard with your zoom calculation. If I understand well, you need to calculate the zoom base on some "paramater" (the distance in km)  that some user set for the markers filter.

If that is correct you could have a helper method in your controller to find the zoom and pass it to the view using the model.

if you need more help, create a sample project with your use case here in GitHub so i can fork it and help you in your scenario

**Comment by net4u** [comment]
Thank you,
I added the missing "using" but now I fall into another errors related to marker.Description (not implemented ?!)
You are very kind. I will prepare a simple test app this weekend. 
Thank you again

**Comment by net4u** [comment]
BTW, I created a fork, but I really don't know how to upload my variation of project

**Comment by jmelosegui** [comment]
Well you should know how to work with github. 
I suggest you [this tutorial](https://www.codeschool.com/courses/mastering-github?gclid=CJL4_8Xcj8UCFWMF7AodVmgAuQ)

**Comment by net4u** [comment]
Hi,

I uploaded a sample project at https://github.com/net4u/GMap-Test hope it helps. I just wonder if is possible to have 2 instance of google map control in same view, each pulling data from different model (e.g. from Jobs and respectively Candidates)

**Comment by net4u** [comment]
In App_Data folder I put the script for database that is behind the models.

**Comment by net4u** [comment]
I finally prepared the support for pulling data from various models (as in my previous question), but still not able to display a map.

**Comment by net4u** [comment]
Hi,

After playing for several days with the control, with success, today I implemented it in a _partialView approach.
After finding a way for successful dealing with multiple models in a single view (via rendering partial views), I made two _PartialViews that I render inside a regular view. The issue that I don't know how to deal with is that the first partial render ok and also display the map, the second don't display the map (but I suppose that somehow it works partially because if I inspect the page the markers are pulled from model and are in their divs. Also I noticed at console level error(s) that say that "You have included the Google Maps API multiple times on this page. This may cause unexpected errors." Since I only referenced google map script only in _Layout I suppose that something is looping.
Also I noticed that I must add :
@section scripts
{
    @(Html.GoogleMap().ScriptRegistrar())
}
also in _partialViews that include the control and also in the view that render them. (I can suppose that from here comes the loop of referencing google map script?!)
Also I noticed that inspecting the view, the divs that hold each map have both id="map" and I suppose that is not ok.
